external-crates/move/move-compiler/src/cfgir/visitor.rs

```diff
@@ -0,0 +1,470 @@
+    fn join(&mut self, other: &Self) -> JoinResult {
+        use LocalState as L;
+        let self_locals = self.locals();
+        let other_locals = other.locals();
+        assert!(
+            self_locals.keys().all(|v| other_locals.contains_key(v)),
+            "ICE. Incorrectly implemented abstract interpreter. \
+            Local variables should not be removed from the map"
+        );
+        assert!(
+            other_locals.keys().all(|v| self_locals.contains_key(v)),
+            "ICE. Incorrectly implemented abstract interpreter. \
+            Local variables should not be removed from the map"
+        );
+        let mut result = JoinResult::Unchanged;
+        for (local, other_state) in other_locals {
+            match (self.locals().get(&local).unwrap(), other_state) {
+                // both available, join the value
+                (L::Available(loc, v1), L::Available(_, v2)) => {
+                    let loc = *loc;
+                    let joined = Self::join_value(v1, v2);
+                    if v1 != &joined {
+                        result = JoinResult::Changed
+                    }
+                    self.locals_mut().insert(*local, L::Available(loc, joined));
+                }
+                // equal so nothing to do
+                (L::Unavailable(_, _), L::Unavailable(_, _))
+                | (L::MaybeUnavailable { .. }, L::MaybeUnavailable { .. }) => (),
+                // if its partially assigned, stays partially assigned
+                (L::MaybeUnavailable { .. }, _) => (),
+
+                // if was partially assigned in other, its now partially assigned
+                (_, L::MaybeUnavailable { .. }) => {
+                    result = JoinResult::Changed;
+                    self.locals_mut().insert(*local, other_state.clone());
+                }
+
+                // Available in one but not the other, so maybe unavailable
+                (L::Available(available, _), L::Unavailable(unavailable, reason))
+                | (L::Unavailable(unavailable, reason), L::Available(available, _)) => {
+                    result = JoinResult::Changed;
+                    let available = *available;
+                    let unavailable = *unavailable;
+                    let state = L::MaybeUnavailable {
+                        available,
+                        unavailable,
+                        unavailable_reason: *reason,
+                    };
+                    self.locals_mut().insert(*local, state);
+                }
+            }
+        }
+        self.join_impl(other, &mut result);
+        result
+    }
+}
+
+/// Trait for simple abstract interpreter passes. Custom hooks can be implemented with additional
+/// logic as needed. The provided implementation will do all of the plumbing of abstract values
+/// through the expressions, commands, and locals.
+pub trait SimpleAbsIntConstructor: Sized {
+    type AI<'a>: Sized + SimpleAbsInt;
```

Comment:
@tzakian, the key ended up being hiding the lifetime in the associated type here. This made it play a lot easier with the `AbstractInterpreterVisitor` trait up above